python/ray/tune/impl/placeholder.py

```diff
@@ -0,0 +1,248 @@
+        }
+    elif isinstance(config, list):
+        return [
+            inject_placeholders(elem, resolvers, id_prefix + (i,), path_prefix + (i,))
+            for i, elem in enumerate(config)
+        ]
+    elif isinstance(config, tuple):
+        return tuple(
+            inject_placeholders(elem, resolvers, id_prefix + (i,), path_prefix + (i,))
+            for i, elem in enumerate(config)
+        )
+    elif _is_primitive(config):
+        # Primitive types.
+        return config
+    elif isinstance(config, Categorical):
+        config.categories = [
+            # Different options gets different id prefixes.
+            # But we should omit appending to path_prefix because after expansion,
+            # this level will not be there.
+            inject_placeholders(choice, resolvers, id_prefix + (i,), path_prefix)
+            for i, choice in enumerate(config.categories)
+        ]
+        return config
+    elif isinstance(config, Function):
+        # Function type.
+        id_hash = _id_hash(id_prefix)
+        v = _FunctionResolver(id_hash, config)
+        resolvers[path_prefix].append(v)
+        return v.get_placeholder()
+    elif not isinstance(config, Domain):
+        # Other non-search space reference objects, dataset, actor handle, etc.
+        id_hash = _id_hash(id_prefix)
+        v = _RefResolver(id_hash, config)
+        resolvers[path_prefix].append(v)
+        return v.get_placeholder()
+    else:
+        # All the other cases, do nothing.
+        return config
+
+
+def _get_placeholder(config: Any, prefix: Tuple, path: Tuple):
+    if not path:
+        return prefix, config
+
+    key = path[0]
+    if (isinstance(config, dict) and key in config) or (
+        isinstance(config, list) and key < len(config)
+    ):
+        # Expand config tree recursively.
+        return _get_placeholder(config[key], prefix=prefix + (path[0],), path=path[1:])
+
+    if isinstance(config, tuple):
+        if config[0] in (_FunctionResolver.TOKEN, _RefResolver.TOKEN):
+            # Found a matching placeholder.
+            # Note that we do not require that the full path are consumed before
+            # declaring a match. Because this placeholder may be part of a nested
+            # search space. For example, the following config:
+            #   config = {
+            #       "param1": tune.grid_search([
+            #           tune.grid_search([Object1, 2, 3]),
+            #           tune.grid_search([Object2, 5, 6]),
+            #       ]),
+            #   }
+            # will result in placeholders under path ("param1", 0, 0).
+            # After expansion though, the choosen placeholder will live under path
+            # ("param1", 0) like this: config = {"param1": (Placeholder1, 2, 3)}
+            return prefix, config
+        elif key < len(config):
+            return _get_placeholder(
+                config[key], prefix=prefix + (path[0],), path=path[1:]
+            )
```

Comment:
Nit: Can we move this regular tuple case up to the first condition?

Something like:

```python
if is_placeholder(config):
    return prefix, config

if list, dict, or tuple:
    recurse
```

Reply:
done